src/lib.rs

```diff
@@ -1,41 +1,386 @@
+        let rectangle_pixel_count = get_pixel_count(rectangle_width, rectangle_height);
+
+        check_painting(&painting, rectangle_pixel_count);
+        paint(
+            self.resolution,
+            &rectangle,
+            rectangle_width,
+            rectangle_height,
+            &mut self.painting,
+            painting,
+        );
+
+        // Payment and NFT minting
+
+        utils::transfer_ftokens(
+            self.ft_program,
+            msg_source,
+            self.owner,
+            rectangle_pixel_count as u128 * self.pixel_price,
+        )
+        .await;
+
+        let token_id = utils::mint_nft(self.nft_program, token_metadata).await;
+        utils::transfer_nft(self.nft_program, msg_source, token_id).await;
+
+        // Insertion and replying
+
+        self.tokens_by_rectangles.insert(
+            rectangle,
+            TokenInfo {
+                owner: msg_source,
+                pixel_price: None,
+                token_id,
+            },
+        );
+        self.rectangles_by_token_ids.insert(token_id, rectangle);
+
+        utils::reply(NFTPixelboardEvent::Minted(token_id));
+    }
+
+    async fn buy(&mut self, token_id: TokenId) {
+        let msg_source = msg::source();
+        let (rectangle, token) = get_mut_token_by_id(
+            &self.rectangles_by_token_ids,
+            &mut self.tokens_by_rectangles,
+            token_id,
+        );
+
+        let pixel_price = token
+            .pixel_price
+            .unwrap_or_else(|| panic!("NFT isn't for sale"));
+        // get_pixel_count() isn't used here because it checks an NFT area for
+        // equality to 0, but here it's always not equal 0.
+        let token_price =
+            (rectangle.width() as usize * rectangle.height() as usize) as u128 * pixel_price;
+        let resale_commission = token_price * self.commission_percentage as u128 / 100;
+
+        utils::transfer_ftokens(self.ft_program, msg_source, self.owner, resale_commission).await;
+        utils::transfer_ftokens(
+            self.ft_program,
+            msg_source,
+            token.owner,
+            token_price - resale_commission,
+        )
+        .await;
+
+        token.pixel_price = None;
+        token.owner = msg_source;
+        utils::transfer_nft(self.nft_program, msg_source, token_id).await;
+
+        utils::reply(NFTPixelboardEvent::Bought(token_id));
+    }
+
+    async fn put_up_for_sale(&mut self, token_id: TokenId, pixel_price: u128) {
+        let (_, token) = get_mut_token_by_id(
+            &self.rectangles_by_token_ids,
+            &mut self.tokens_by_rectangles,
+            token_id,
+        );
+        assert_eq!(token.owner, msg::source());
+        assert!(token.pixel_price.is_none(), "NFT is already for sale");
+
+        check_pixel_price(pixel_price);
+        token.pixel_price = Some(pixel_price);
+        utils::transfer_nft(self.nft_program, exec::program_id(), token_id).await;
```

Comment:
Should we transfer NFT to the program account? NFT can be approved and be in the user wallet until sale